examples/http_async_server.hpp

```diff
@@ -67,13 +71,138 @@ class http_async_server
             t.join();
     }
 
+    void
+    log_args()
+    {
+    }
+
+    template<class Arg, class... Args>
+    void
+    log_args(Arg const& arg, Args const&... args)
+    {
+        std::cerr << arg;
+        log_args(args...);
+    }
+
+    template<class... Args>
+    void
+    log(Args const&... args)
+    {
+        if(log_)
+        {
+            std::lock_guard<std::mutex> lock(m_);
+            log_args(args...);
+            std::cerr.flush();
+        }
+    }
+
 private:
+    template<class Stream, class Handler,
+        bool isRequest, class Body, class Headers>
+    class write_op
+    {
+        using alloc_type =
+            handler_alloc<char, Handler>;
+
+        struct data
+        {
+            Stream& s;
+            message_v1<isRequest, Body, Headers> m;
+            Handler h;
+            bool cont;
+
+            template<class DeducedHandler>
+            data(DeducedHandler&& h_, Stream& s_,
+                    message_v1<isRequest, Body, Headers>&& m_)
+                : s(s_)
+                , m(std::move(m_))
+                , h(std::forward<DeducedHandler>(h_))
+                , cont(boost_asio_handler_cont_helpers::
+                    is_continuation(h))
+            {
+            }
+        };
+
+        std::shared_ptr<data> d_;
+
+    public:
+        write_op(write_op&&) = default;
+        write_op(write_op const&) = default;
+
+        template<class DeducedHandler, class... Args>
+        write_op(DeducedHandler&& h, Stream& s, Args&&... args)
+            : d_(std::allocate_shared<data>(alloc_type{h},
+                std::forward<DeducedHandler>(h), s,
+                    std::forward<Args>(args)...))
+        {
+            (*this)(error_code{}, false);
+        }
+
+        void
+        operator()(error_code ec, bool again = true)
+        {
+            auto& d = *d_;
+            d.cont = d.cont || again;
+            if(! again)
+            {
+                beast::http::async_write(d.s, d.m, std::move(*this));
+                return;
+            }
+            d.h(ec);
+        }
+
+        friend
+        void* asio_handler_allocate(
+            std::size_t size, write_op* op)
+        {
+            return boost_asio_handler_alloc_helpers::
+                allocate(size, op->d_->h);
+        }
+
+        friend
+        void asio_handler_deallocate(
+            void* p, std::size_t size, write_op* op)
+        {
+            return boost_asio_handler_alloc_helpers::
+                deallocate(p, size, op->d_->h);
+        }
+
+        friend
+        bool asio_handler_is_continuation(write_op* op)
+        {
+            return op->d_->cont;
+        }
+
+        template <class Function>
+        friend
+        void asio_handler_invoke(Function&& f, write_op* op)
+        {
+            return boost_asio_handler_invoke_helpers::
+                invoke(f, op->d_->h);
```

Comment:
Definitely not. The intent here is that `f` can be passed as either `Function const&` or `Function&`.